app/services/activitypub/synchronize_followers_service.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+class ActivityPub::SynchronizeFollowersService < BaseService
+  include JsonLdHelper
+  include Payloadable
+
+  def call(account, partial_collection_url)
+    @account = account
+
+    items = collection_items(partial_collection_url)
+    return if items.nil?
+
+    @expected_followers = items.map { |uri| ActivityPub::TagManager.instance.uri_to_resource(uri, Account) }.compact
+    # TODO: what to do with unresolvable accounts?
+
+    remove_unexpected_local_followers!
+    undo_unexpected_outgoing_follows!
+  end
+
+  private
+
+  def remove_unexpected_local_followers!
+    @account.followers.local.where.not(id: @expected_followers.map(&:id)).each do |unexpected_follower|
+      UnfollowService.new.call(unexpected_follower, @account)
+    end
+  end
+
+  def undo_unexpected_outgoing_follows!
+    @expected_followers.each do |expected_follower|
+      next if expected_follower.following?(@account)
+
+      follow = Follow.new(account: expected_follower, target_account: @account)
+      # TODO: the follow doesn't have an uri, how does that work?
```

Comment:
Indeed, this is awkward. But I'm not sure why this is a TODO and a question. Perhaps change it to a comment that warns:

> Since we were not aware of the follow from our side, we do not have an ID for it that we can include in the Undo activity. For this reason, the Undo may not work with software that relies exclusively on matching activity IDs and not the actor and target